pip/pip-281.md

```diff
@@ -0,0 +1,233 @@
+* Avoid code snippets, unless it's essential to explain your intent.
+-->
+
+Current Unload and Lookup Sequence in Extensible Load Balancer
+```mermaid
+sequenceDiagram
+    participant Clients
+    participant Owner Broker
+    participant New Owner Broker
+    participant Leader Broker
+    Leader Broker ->> Owner Broker: "state:Releasing:" close topic
+    Owner Broker ->> Owner Broker: close broker topic sessions
+    Owner Broker ->> Clients: close producers and consumers
+    Clients ->> Clients: reconnecting (inital delay 100ms)
+    Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
+    New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
+    Clients ->> Owner Broker: lookup
+    Owner Broker ->> Clients: redirect
+    Clients ->> New Owner Broker: lookup
+    New Owner Broker ->> Clients: return(connected)
+```
+
+Proposed Unload Sequence in Extensible Load Balancer without Lookup
+```mermaid
+sequenceDiagram
+    participant Clients
+    participant Owner Broker
+    participant New Owner Broker
+    participant Leader Broker
+    Leader Broker ->> Owner Broker: "state:Releasing:" close topic
+    Owner Broker ->> Owner Broker: close broker topic sessions(e.g ledgers) without disconnecting producers/consumers(fenced)
+    Clients -->> Owner Broker: message pubs are ignored
+    Owner Broker ->> New Owner Broker: "state:Assign:" assign new ownership
+    New Owner Broker ->> Owner Broker: "state:Owned:" ack new ownership
+    Owner Broker ->> Owner Broker: close the fenced broker topic sessions
+    Owner Broker ->> Clients: close producers and consumers (with newOwnerBrokerUrl)
+    Clients ->> New Owner Broker: immediately connect
+```
+
+
+# Detailed Design
+
+## Design & Implementation Details
+
+<!--
+This is the section where you dive into the details. It can be:
+* Concrete class names and their roles and responsibility, including methods.
+* Code snippets of existing code.
+* Interface names and its methods.
+* ...
+-->
+
+- Modify CommandCloseProducer, CommandCloseConsumer to pass optional brokerServiceUrls
+```
+message CommandCloseProducer {
+required uint64 producer_id = 1;
+required uint64 request_id = 2;
++ optional string assignedBrokerServiceUrl      = 3;
++ optional string assignedBrokerServiceUrlTls   = 4;
+}
+
+message CommandCloseConsumer {
+required uint64 consumer_id = 1;
+required uint64 request_id = 2;
++ optional string assignedBrokerServiceUrl      = 3;
++ optional string assignedBrokerServiceUrlTls   = 4;
+}
+```
+
+- Add new disconnect apis on producer and consumer to pass dstBrokerLookupData
+```
+public CompletableFuture<Void> disconnect(Optional<BrokerLookupData> dstBrokerLookupData) {
+```
+
+- Modify the Topic.close() behavior to optionally skip producers.disconnect() and consumers.disconnect().
+```
+public CompletableFuture<Void> close(boolean closeWithoutWaitingClientDisconnect,
+                                          boolean closeWithoutDisconnectingClients) {
```

Comment:
Sorry, I'm not very clear about the option `closeWithoutDisconnectingClients`, could you add some explanations?

Reply:
As I described in the sequence diagram, this `closeWithoutDisconnectingClients ` flag controls disconnecting clients upon closing topics.

If closeWithoutDisconnectingClients=false, this is the current behavior. It disconnects clients(producers and consumers)

If closeWithoutDisconnectingClients=true, it will close the topic session(e.g. ledger) without disconnecting clients(producers and consumers).

From the transfer protocol perspective, releasing->assigned->owned state transitions, 
, with this change,
at the releasing state, the source broker will close the topic with closeWithoutDisconnectingClients=true.
at the owned state, the source broker will close the topic with closeWithoutDisconnectingClients=false.

With this flow control, we can gracefully close the ledgers before the destination broker acquires the bundle ownership. This will result in fewer ledger creation conflicts during the unload(graceful bundle ownership transfer).

Reply:
Oh, if we close clients in releasing state, clients will still take the lookup operation, right?

> This will result in fewer ledger creation conflicts during the unload(graceful bundle ownership transfer).

Sorry, I'm not clear about this, why there are ledger conflicts during the unload?

Reply:
> Oh, if we close clients in releasing state, clients will still take the lookup operation, right?
As the sequence diagram describes, we don't close(disconnect) clients at the releasing state. We disconnect the clients at the owned state.


>> This will result in fewer ledger creation conflicts during the unload(graceful bundle ownership transfer).
> Sorry, I'm not clear about this, why there are ledger conflicts during the unload?

Although we have 100 ms delay upon clients' reconnection, I think there is still a racing condition between reconnection(ledger creation) and ledger close during unloading. I believe this PIP can gracefully synchronize this ledger close and reconnection(ledger creation) sequence by separating them into different states(releasing and owned).

Reply:
Ok, thanks for the explanation.

Reply:
> Although we have 100 ms delay upon clients' reconnection, I think there is still a racing condition between reconnection(ledger creation) and ledger close during unloading. I believe this PIP can help gracefully synchronize this ledger close and reconnection(ledger creation) sequence by separating them into different states(releasing and owned).

But a message from the connected producer will also cause the producers and consumers to disconnect.

https://github.com/apache/pulsar/blob/ee91edc3e08c87db1e5662a553ff62af0c1886e5/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L640-L682